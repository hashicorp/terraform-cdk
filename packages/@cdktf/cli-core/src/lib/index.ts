// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
// This is the programmatic entrypoint that the CLI uses.
// While this is the closest we have to a programmatic API, please understand that the interfaces in this file are not stable.
// Convert is not included here since it's published independently as @cdktf/hcl2cdk.
/* eslint-disable @typescript-eslint/no-non-null-assertion */

export {
  DependencyManager,
  ProviderConstraint,
} from "./dependencies/dependency-manager";
export { getAllPrebuiltProviders } from "./dependencies/prebuilt-providers";
export { providerAdd } from "./provider-add";
export { TerraformCli } from "./models/terraform-cli";
export { CdktfConfig } from "./cdktf-config";
export { initializErrorReporting } from "./error-reporting";
export { CdktfStack } from "./cdktf-stack";
export { init, Project, InitArgs, templates, templatesDir } from "./init";
export { get, GetStatus, runGetInDir } from "./get";
export { SynthesizedStack } from "./synth-stack";
export { CdktfProject, ProjectUpdate } from "./cdktf-project";
export { watch, State as WatchState } from "./watch";
export {
  NestedTerraformOutputs,
  saveOutputs,
  normalizeOutputPath,
} from "./output";
export {
  DeployingResource,
  DeployingResourceApplyState,
  isTerraformOutput,
  PlannedResourceAction,
  TerraformOutput,
} from "./models/terraform";
