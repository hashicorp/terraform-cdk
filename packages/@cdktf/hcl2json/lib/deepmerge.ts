// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
/**
 * Merges `source` into `target`, overriding any existing values.
 * `undefined` will cause a value to be deleted.
 */
export function deepMerge(target: any, ...sources: any[]) {
  for (const source of sources) {
    if (typeof source !== "object" || typeof target !== "object") {
      throw new Error(
        `Invalid usage. Both source (${JSON.stringify(
          source
        )}) and target (${JSON.stringify(target)}) must be objects`
      );
    }

    for (const key of Object.keys(source)) {
      const value = source[key];
      if (typeof value === "object" && value != null && !Array.isArray(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (typeof target[key] !== "object") {
          target[key] = {};
        }

        deepMerge(target[key], value);

        // if the result of the merge is an empty object, it's because the
        // eventual value we assigned is `undefined`, and there are no
        // sibling concrete values alongside, so we can delete this tree.
        const output = target[key];
        if (typeof output === "object" && Object.keys(output).length === 0) {
          delete target[key];
        }
      } else if (
        typeof value === "object" &&
        value != null &&
        Array.isArray(value)
      ) {
        if (Array.isArray(target[key])) {
          target[key] = [...target[key], ...value];
        } else {
          target[key] = value;
        }
      } else if (value === undefined) {
        delete target[key];
      } else {
        target[key] = value;
      }
    }
  }

  return target;
}
