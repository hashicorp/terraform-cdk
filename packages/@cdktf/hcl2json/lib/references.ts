// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0

import {
  ExpressionAst,
  ScopeTraversalExpressionMeta,
  TerraformTraversalPart,
  ForExpressionMeta,
  getChildWithValue,
} from "./syntax-tree";

export type Reference = {
  value: string;
  startPosition: number;
  endPosition: number;
};

export function traversalToReference(
  traversalExpression: ExpressionAst,
  localVariables?: string[],
): Reference | null {
  const meta = traversalExpression.meta as ScopeTraversalExpressionMeta;

  // We do not want to include property access through brackets here
  // although it is technically a traversal / reference
  function onlyTakeTraversalPartsUntilFirstBracketPropertyAccess(
    traversals: TerraformTraversalPart[],
  ) {
    let filtered = [];
    let index = 0;

    for (const traversal of traversals) {
      if ("segment" in traversal) {
        if (index === 0) {
          // We are at the first traversal, check if it is a local variable
          if (localVariables?.includes(traversal.segment)) {
            // We reached a local variable, stop
            return [];
          }
        }
        filtered.push(traversal);
      } else {
        // We reached a bracket, stop
        return filtered;
      }
    }

    return filtered;
  }

  const filteredParts = onlyTakeTraversalPartsUntilFirstBracketPropertyAccess(
    meta.traversal,
  );

  if (filteredParts.length === 0) {
    return null;
  }

  const startPosition = filteredParts[0].range.start.byte;
  const endPosition = filteredParts[filteredParts.length - 1].range.end.byte;

  return {
    value: filteredParts.map((part) => part.segment).join("."),
    startPosition,
    endPosition,
  };
}

export function findAllReferencesInAst(
  input: string,
  entry: ExpressionAst | undefined | null,
  localVariables?: string[],
): Reference[] {
  if (!entry) {
    return [];
  }

  switch (entry.type) {
    case "scopeTraversal": {
      // For traversals within a for expression,
      // we want to ignore the local variables of the for expression
      const reference = traversalToReference(entry, localVariables);
      if (reference) return [reference];
      return [];
    }

    case "for": {
      const meta = entry.meta as ForExpressionMeta;
      const additionalLocalVariables = [
        meta.keyVar,
        meta.valVar,
        ...(localVariables || []),
      ];

      return [
        ...findAllReferencesInAst(
          input,
          getChildWithValue(entry, meta.collectionExpression),
          additionalLocalVariables,
        ),
        ...findAllReferencesInAst(
          input,
          getChildWithValue(entry, meta.conditionalExpression),
          additionalLocalVariables,
        ),
      ];
    }
    default:
      return entry.children
        .map((child) => findAllReferencesInAst(input, child, localVariables))
        .flat();
  }
}
