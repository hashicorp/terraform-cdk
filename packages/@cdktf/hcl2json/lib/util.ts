/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

/**
 * This function is used to escape quotes within a string.
 *  Note: This is a really naive implmentation that will not be able to work with
 *    nested templates. However, nested templates require parsing, so we will
 *    only handle strings without nested templates for the moment
 * @param input The string to escape
 */
export function replaceQuotes(input: string): string {
  const fullyQuotedMatch = /^(\s*")(.*)("\s*)$/.exec(input);
  const hasTemplate = /(?<!\$)\$\{/.test(input);

  if (fullyQuotedMatch !== null && !hasTemplate) {
    return `${fullyQuotedMatch[1]}${fullyQuotedMatch[2].replace(/"/g, `\\"`)}${
      fullyQuotedMatch[3]
    }`;
  }

  return input;
}

/**
 * This function is used to wrap a string in quotes conditionally.
 *  - If the string is already quoted, it will be returned as-is.
 *  - If the string contains a newline, it will be wrapped in a heredoc.
 *  - If the string is a number, it will be returned as-is.
 *  - If the string is a boolean, it will be returned as-is.
 *  - If the string is a list or map, it will be returned as-is.
 *
 * @param input The string to wrap
 * @returns The wrapped string, and the offset of the cursor within the string
 *
 * @example
 * wrapTerraformExpression("foo") // { wrap: '"foo"', wrapOffset: 1 }
 * wrapTerraformExpression("foo\nbar") // { wrap: '<<EOE\nfoo\nbar\nEOE\n', wrapOffset: 5 }
 * wrapTerraformExpression("22") // { wrap: '22', wrapOffset: 0 }
 * wrapTerraformExpression("true") // { wrap: 'true', wrapOffset: 0 }
 * wrapTerraformExpression("[22, 10]") // { wrap: '[22, 10]', wrapOffset: 0 }
 * wrapTerraformExpression("{ foo = 22 }") // { wrap: '{ foo = 22 }', wrapOffset: 0 }
 */
export function wrapTerraformExpression(input: string): {
  wrap: string;
  wrapOffset: number;
} {
  input = replaceQuotes(input);

  if (!isNaN(Number(input))) {
    return { wrap: input, wrapOffset: 0 };
  }
  if (input === "true" || input === "false") {
    return { wrap: input, wrapOffset: 0 };
  }

  if (input.startsWith(`"`)) {
    if (input.indexOf("\n") >= 0) {
      const trimWrapped = input.substring(1, input.length - 1);
      return {
        wrap: `<<EOE\n${trimWrapped}\nEOE\n`,
        wrapOffset: 5,
      };
    }
    return { wrap: input, wrapOffset: 0 };
  }

  if (input.startsWith("[") || input.startsWith("{")) {
    return { wrap: input, wrapOffset: 0 };
  }

  if (input.startsWith("<<")) {
    // For Heredocs, we need to ensure that the string ends with an empty newline as
    // the HCL parser doesn't find the ending token otherwise
    if (!input.endsWith("\n")) {
      return { wrap: input + "\n", wrapOffset: 0 };
    }

    return { wrap: input, wrapOffset: 0 };
  }

  return { wrap: `"${input}"`, wrapOffset: 1 };
}
