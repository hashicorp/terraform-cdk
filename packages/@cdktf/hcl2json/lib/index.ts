// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference lib="dom" />

// Inspired by
// https://github.com/ts-terraform/ts-terraform
// https://github.com/aaronpowell/webpack-golang-wasm-async-loader

import fs from "fs-extra";
import path from "path";
import { deepMerge } from "./deepmerge";
import { gunzipSync } from "zlib";

interface GoBridge {
  parse: (filename: string, hcl: string) => Promise<string>;
  getReferencesInExpression: (filename: string, hcl: string) => Promise<string>;
}

// eslint-disable-next-line @typescript-eslint/ban-types
const jsRoot: Record<string, Function> = {};

function sleep() {
  return new Promise((resolve) => {
    setTimeout(resolve, 0);
  });
}

function goBridge(getBytes: Promise<Buffer>) {
  let ready = false;

  async function init() {
    // After: https://github.com/golang/go/commit/680caf15355057ca84857a2a291b6f5c44e73329
    // Go 1.19+ has a different entrypoint
    await import(`../wasm/bridge_wasm_exec.js`);
    const go = new (global as any).Go();
    const bytes = await getBytes;
    const result = await WebAssembly.instantiate(bytes, go.importObject);
    (global as any).__parse_terraform_config_wasm__ = jsRoot;
    void go.run(result.instance);
    ready = true;
  }

  init().catch((error) => {
    throw error;
  });

  const proxy = new Proxy({} as GoBridge, {
    get: (_, key: string) => {
      return async (...args: unknown[]) => {
        while (!ready) {
          await sleep();
        }

        if (!(key in jsRoot)) {
          throw new Error(
            `There is nothing defined with the name "${key.toString()}"`
          );
        }

        if (typeof jsRoot[key] !== "function") {
          return jsRoot[key];
        }

        return new Promise((resolve, reject) => {
          const cb = (err: string, ...msg: string[]) =>
            // @ts-ignore
            err ? reject(new Error(err)) : resolve(...msg);

          const run = () => {
            jsRoot[key].apply(undefined, [...args, cb]);
          };

          run();
        });
      };
    },
  });

  return proxy;
}

const loadWasm = async () => {
  return gunzipSync(
    await fs.readFile(path.join(__dirname, "..", "main.wasm.gz"))
  );
};

const wasm = goBridge(loadWasm());

export async function parse(
  filename: string,
  contents: string
): Promise<Record<string, any>> {
  const res = await wasm.parse(filename, contents);
  return JSON.parse(res);
}

export async function convertFiles(
  workingDirectory: string
): Promise<Record<string, any> | void> {
  let tfFileContents = "";
  const tfJSONFileContents: Record<string, any>[] = [];

  for (const file of fs.readdirSync(workingDirectory)) {
    const filePath = path.resolve(workingDirectory, file);
    if (!fs.lstatSync(filePath).isDirectory()) {
      if (file.match(/\.tf$/)) {
        tfFileContents += fs.readFileSync(filePath, "utf-8");
        tfFileContents += "\n";
      } else if (file.match(/\.tf\.json$/)) {
        tfJSONFileContents.push(JSON.parse(fs.readFileSync(filePath, "utf-8")));
      }
    }
  }

  if (tfFileContents === "" && tfJSONFileContents === []) {
    console.error(`No '.tf' or '.tf.json' files found in ${workingDirectory}`);
    return;
  }

  return deepMerge(
    await parse("hcl2json.tf", tfFileContents),
    ...tfJSONFileContents
  );
}

type CodeMarker = {
  Byte: number;
  Line: number;
  Column: number;
};
type Range = {
  End: CodeMarker;
  Start: CodeMarker;
};
type TerraformTraversalPart = {
  Name: string;
  SrcRange: Range;
};
// Reference to a variable / module / resource
type TerraformTraversal = { Traversal: TerraformTraversalPart[] };
type PropertyAccessExpression = TerraformTraversal & {
  Source: TerraformObject; // it's probably more restricted, but in this context we don't care
};
type TerraformFunctionCall = {
  Args: TerraformObject[];
  Name: string;
  ExpandFinal: boolean;
  NameRange: Range;
  OpenParenRange: Range;
  CloseParenRange: Range;
};
type TerraformLiteral = {
  SrcRange: Range;
  Val: unknown; // No value is passed down, we ignore them
};
type TerraformEmbeddedExpression = {
  Wrapped: TerraformObject;
};
type TerraformExpression = {
  Parts: TerraformObject[];
};
type ArithmeticExpression = {
  LHS: TerraformObject;
  RHS: TerraformObject;
};
type ConditionExpression = {
  Condition: TerraformObject;
  TrueResult: TerraformObject;
  FalseResult: TerraformObject;
};
type ForExpression = {
  KeyVar: string;
  ValVar: string;
  CollExpr: TerraformObject;
  KeyExpr: TerraformObject;
  ValExpr: TerraformObject;
  CondExpr: TerraformObject;
};
type BracketPropertyAccessExpression = {
  Collection: TerraformObject;
  Key: TerraformObject; // it's probably more restricted, but in this context we don't care
};
type ListExpression = {
  Exprs?: TerraformObject[] | null;
};

type TerraformObject =
  | ArithmeticExpression
  | BracketPropertyAccessExpression
  | ConditionExpression
  | ForExpression
  | ListExpression
  | PropertyAccessExpression
  | TerraformEmbeddedExpression
  | TerraformExpression
  | TerraformFunctionCall
  | TerraformLiteral
  | TerraformTraversal;

type GoExpressionParseResult = null | TerraformObject;

type Reference = { value: string; startPosition: number; endPosition: number };

function traversalToReference(
  input: string,
  traversal: TerraformTraversal
): Reference {
  const lines = input.split("\n");
  const lineLength = lines.map((line) => line.length);

  function position(marker: CodeMarker) {
    const newlineChar = 1;
    return (
      lineLength
        .slice(0, marker.Line)
        .reduce((a, b) => a + b + newlineChar, lines.length === 1 ? 0 : -1) +
      marker.Column
    );
  }

  // We do not want to include property access through brackets here
  // although it is technically a traversal / reference
  function onlyTakeTraversalPartsUntilFirstBracketPropertyAccess(
    traversals: TerraformTraversalPart[]
  ) {
    let filtered = [];

    for (const traversal of traversals) {
      if ("Name" in traversal) {
        filtered.push(traversal);
      } else {
        // We reached a bracket, stop
        return filtered;
      }
    }

    return filtered;
  }
  const filteredParts = onlyTakeTraversalPartsUntilFirstBracketPropertyAccess(
    traversal.Traversal
  );

  const startPosition = position(filteredParts[0].SrcRange.Start);
  const endPosition = position(
    filteredParts[filteredParts.length - 1].SrcRange.End
  );

  return {
    value: input.slice(startPosition, endPosition),
    startPosition,
    endPosition,
  };
}

function findAllReferencesInAst(
  input: string,
  entry: TerraformObject | undefined | null
): Reference[] {
  if (!entry) {
    return [];
  }

  // This traversal is accessing a function call e.g. element(var.foo, 0).id
  // Or this is a splat operation var.foo.*
  if ("Source" in entry) {
    return findAllReferencesInAst(input, entry.Source);
  }

  // a.b.c.d is a traversal
  if ("Traversal" in entry) {
    return [traversalToReference(input, entry)];
  }

  // Multiple terraform parts in an expression
  if ("Parts" in entry) {
    return entry.Parts.flatMap((part) => findAllReferencesInAst(input, part));
  }

  // ${} is an embedded expression
  if ("Wrapped" in entry) {
    return findAllReferencesInAst(input, entry.Wrapped);
  }

  // element(var.foo, 0) is a function call
  if ("Args" in entry) {
    return entry.Args.flatMap((arg) => findAllReferencesInAst(input, arg));
  }

  // var.foo + var.bar is an arithmetic expression
  if ("LHS" in entry) {
    return [
      ...findAllReferencesInAst(input, entry.LHS),
      ...findAllReferencesInAst(input, entry.RHS),
    ];
  }

  // var.foo > 3 ? var.bar : var.baz is a condition expression
  if ("Condition" in entry) {
    return [
      ...findAllReferencesInAst(input, entry.Condition),
      ...findAllReferencesInAst(input, entry.TrueResult),
      ...findAllReferencesInAst(input, entry.FalseResult),
    ];
  }

  // for name, user in var.users : user.role => name... is a for expression
  if ("CondExpr" in entry) {
    return [
      ...findAllReferencesInAst(input, entry.CollExpr),
      ...findAllReferencesInAst(input, entry.CondExpr),
    ];
  }

  // var.foo["bar"] is a bracket property access expression
  if ("Key" in entry) {
    return [
      ...findAllReferencesInAst(input, entry.Collection),
      ...findAllReferencesInAst(input, entry.Key),
    ];
  }

  // [var.foo, var.bar] is a list expression
  if ("Exprs" in entry) {
    return (entry.Exprs || []).flatMap((expr) =>
      findAllReferencesInAst(input, expr)
    );
  }

  return [];
}

export async function getReferencesInExpression(
  filename: string,
  expression: string
): Promise<Reference[]> {
  const res = await wasm.getReferencesInExpression(
    filename,
    JSON.stringify(expression)
  );
  const ast = JSON.parse(res) as GoExpressionParseResult;

  if (!ast) {
    return [];
  }

  return findAllReferencesInAst(expression, ast);
}
