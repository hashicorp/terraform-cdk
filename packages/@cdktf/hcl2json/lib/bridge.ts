// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference lib="dom" />

// Inspired by
// https://github.com/ts-terraform/ts-terraform
// https://github.com/aaronpowell/webpack-golang-wasm-async-loader

import fs from "fs-extra";
import path from "path";
import { deepMerge } from "./deepmerge";
import { gunzipSync } from "zlib";
import { Reference, findAllReferencesInAst } from "./references";
import { ExpressionType } from "./syntax-tree";
import { wrapTerraformExpression } from "./util";

interface GoBridge {
  parse: (filename: string, hcl: string) => Promise<string>;
  parseExpression: (filename: string, hcl: string) => Promise<string>;
  getExpressionAst: (filename: string, hcl: string) => Promise<string>;
}

// eslint-disable-next-line @typescript-eslint/ban-types
const jsRoot: Record<string, Function> = {};

function sleep() {
  return new Promise((resolve) => {
    setTimeout(resolve, 0);
  });
}

function goBridge(getBytes: Promise<Buffer>) {
  let ready = false;

  async function init() {
    // After: https://github.com/golang/go/commit/680caf15355057ca84857a2a291b6f5c44e73329
    // Go 1.19+ has a different entrypoint
    await import(`../wasm/bridge_wasm_exec.js`);
    const go = new (global as any).Go();
    const bytes = await getBytes;
    const result = await WebAssembly.instantiate(bytes, go.importObject);
    (global as any).__parse_terraform_config_wasm__ = jsRoot;
    void go.run(result.instance);
    ready = true;
  }

  init().catch((error) => {
    throw error;
  });

  const proxy = new Proxy({} as GoBridge, {
    get: (_, key: string) => {
      return async (...args: unknown[]) => {
        while (!ready) {
          await sleep();
        }

        if (!(key in jsRoot)) {
          throw new Error(
            `There is nothing defined with the name "${key.toString()}"`
          );
        }

        if (typeof jsRoot[key] !== "function") {
          return jsRoot[key];
        }

        return new Promise((resolve, reject) => {
          const cb = (err: string, ...msg: string[]) =>
            // @ts-ignore
            err ? reject(new Error(err)) : resolve(...msg);

          const run = () => {
            jsRoot[key].apply(undefined, [...args, cb]);
          };

          run();
        });
      };
    },
  });

  return proxy;
}

const loadWasm = async () => {
  return gunzipSync(
    await fs.readFile(path.join(__dirname, "..", "main.wasm.gz"))
  );
};

const wasm = goBridge(loadWasm());

export async function parse(
  filename: string,
  contents: string
): Promise<Record<string, any>> {
  const res = await wasm.parse(filename, contents);
  return JSON.parse(res);
}

export async function convertFiles(
  workingDirectory: string
): Promise<Record<string, any>> {
  let tfFileContents = "";
  const tfJSONFileContents: Record<string, any>[] = [];

  for (const file of fs.readdirSync(workingDirectory)) {
    const filePath = path.resolve(workingDirectory, file);
    if (!fs.lstatSync(filePath).isDirectory()) {
      if (file.match(/\.tf$/)) {
        tfFileContents += fs.readFileSync(filePath, "utf-8");
        tfFileContents += "\n";
      } else if (file.match(/\.tf\.json$/)) {
        tfJSONFileContents.push(JSON.parse(fs.readFileSync(filePath, "utf-8")));
      }
    }
  }

  if (tfFileContents === "" && tfJSONFileContents.length === 0) {
    console.error(`No '.tf' or '.tf.json' files found in ${workingDirectory}`);
    return {};
  }

  return deepMerge(
    await parse("hcl2json.tf", tfFileContents),
    ...tfJSONFileContents
  );
}

/**
 * Parse a Terraform expression and return the AST. This function expects a string input, and will wrap the expression in quotes if it is not already quoted.
 * @param filename The filename to use for the expression. This is used for error reporting.
 * @param expression The expression to parse.
 * @returns An array of References found in the expression.
 */
export async function getReferencesInExpression(
  filename: string,
  expression: string
): Promise<Reference[]> {
  // We have to do this twice because of the problem with HEREDOCS
  // Our current hcl2json implementation removes HEREDOCS and replaces them
  // with a multi-line string, which is causing all kinds of problems
  let offset = 0;
  let quoteWrappedExpression = expression;
  if (!expression.startsWith('"')) {
    quoteWrappedExpression = `"${expression}"`;
    offset = 1;
  }

  const { wrap: wrappedExpression, wrapOffset: startOffset } =
    wrapTerraformExpression(`${quoteWrappedExpression}`);

  offset += startOffset;

  const ast = await getExpressionAst(filename, wrappedExpression);
  if (!ast) {
    return [];
  }

  const refs = findAllReferencesInAst(expression, ast);
  if (wrappedExpression === expression) {
    return refs;
  }

  return refs.map((ref) => {
    return {
      ...ref,
      startPosition: ref.startPosition - offset,
      endPosition: ref.endPosition - offset,
    };
  });
}

/**
 * Parse a Terraform expression and return the AST. The expression does not need to be a Terraform string.
 * @param filename The filename to use for the expression. This is used for error reporting.
 * @param expression The expression to parse.
 * @returns The AST for the expression.
 *
 *   The returned AST has the following structure:
 *   - type: The type of the node. This is a string.
 *   - range: This contains the start and end of the node in the expression.
 *   - children: This contains the children of the node. This is an array of nodes.
 *   - meta: This contains metadata about the node. This is an object, and varies depending on the type of the node.
 */
export async function getExpressionAst(
  filename: string,
  expression: string
): Promise<ExpressionType | null> {
  const res = await wasm.getExpressionAst(filename, expression);
  const ast = JSON.parse(res) as ExpressionType;

  if (!ast) {
    return null;
  }

  return ast;
}
