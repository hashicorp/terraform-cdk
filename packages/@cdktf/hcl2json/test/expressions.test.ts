// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { getReferencesInExpression, getExpressionAst } from "../lib";

describe("getReferencesInExpression", () => {
  test.each([
    ["parses a simple string", '"foo"'],
    ["parses a reference string", '"${var.foo}"'],
    ["parses three part references", '"${module.foo.output}"'],
    ["parses four part references", '"${data.type.name.attr}"'],
    [
      "parses a terraform function",
      '"${replace(module.foo.output, "-", var.bar)}"',
    ],
    [
      "parses nested terraform functions",
      '"${split(var.separator, lower(replace(module.foo.output, "-", var.bar)))}"',
    ],
    [
      "parses embedded vars",
      '"Hey, did you hear about ${module.foo.output}s new album?"',
    ],
    ["parses lists", '"${[5, "${module.foo.output}", "val", true]}"'],
    [
      "parses maps",
      '"${{ x = "${module.foo.output}", y = "val", z = true, a = "${replace(module.foo.output, "-", var.bar)}" }}"',
    ],
    ["parses variable arithmetics", '"${var.members + var.admins}"'],
    [
      "parses ternary expressions",
      '"${aws_kms_key.examplekms.deletion_window_in_days > 3 ? aws_s3_bucket.examplebucket.id : []}"',
    ],
    [
      "parses multi-line for loops",
      '"${{\n            for name, user in var.users : name => user\n            if user.is_admin\n          }}"',
    ],
    [
      "parses for in loops",
      '"${{ for name, user in var.users : user.role => name... }}"',
    ],
    [
      "parses terraform functions with file access",
      '"${ element(aws_s3_bucket.examplebucket_two, 0).id }"',
    ],
    ["parses splat expressions", '"${ toset(aws_s3_bucket.examplebucket.*) }"'],
    ["property access", `"\${ var.setting["bucket_name"] }"`],
    ["mixed property access", `"\${ var.setting["bucket_name"].arn }"`],
    ["variable property access", `"\${ var.settings[var.bucket_name_key] }"`],
    [
      "partial variable property access",
      `"\${ var.settings["\${var.prefix}-my-bucket"] }"`,
    ],
    ["list expressions", `"\${ [var.tags.app, var.tags.env] }"`],
  ])("%s", async (_, input) => {
    const res = await getReferencesInExpression("main.tf", input);

    expect(res).toMatchSnapshot();
  });

  test("selects the right parts in multi-line expressions", async () => {
    const input = `\${merge(
      var.settings,
      var.other_settings, # best one
      var.yet_another_setting,
      var.final_setting,
    )}`;
    const references = await getReferencesInExpression("main.tf", input);

    return expect(
      references.map((ref) =>
        input.substring(ref.startPosition, ref.endPosition)
      )
    ).toMatchInlineSnapshot(`
              [
                "var.settings",
                "var.other_settings",
                "var.yet_another_setting",
                "var.final_setting",
              ]
            `);
  });

  test("fails on malformed expressions", () => {
    return expect(
      getReferencesInExpression("main.tf", '"${module.foo.output"')
    ).rejects.toMatchSnapshot();
  });
});

describe("getExpressionAst", () => {
  test.each([
    ["parses a simple string", "foo"],
    ["parses a reference string", '"${var.foo}"'],
    ["parses three part references", '"${module.foo.output}"'],
    ["parses four part references", '"${data.type.name.attr}"'],
    [
      "parses a terraform function",
      '"${replace(module.foo.output, "-", var.bar)}"',
    ],
    [
      "parses nested terraform functions",
      '"${split(var.separator, lower(replace(module.foo.output, "-", var.bar)))}"',
    ],
    [
      "parses embedded vars",
      '"Hey, did you hear about ${module.foo.output}s new album?"',
    ],
    ["parses lists", '[5, "${module.foo.output}", "val", true]'],
    [
      "parses maps",
      '{ x = "${module.foo.output}", y = "val", z = true, a = "${replace(module.foo.output, "-", var.bar)}" }',
    ],
    ["parses variable arithmetics", '"${var.members + var.admins}"'],
    [
      "parses ternary expressions",
      '"${aws_kms_key.examplekms.deletion_window_in_days > 3 ? aws_s3_bucket.examplebucket.id : []}"',
    ],
    [
      "parses multi-line for loops",
      '"${{\n            for name, user in var.users : name => user\n            if !user.is_admin\n          }}"',
    ],
    [
      "parses for in loops",
      '"${{ for name, user in var.users : user.role => name... }}"',
    ],
    [
      "parses terraform functions with file access",
      '"${ element(aws_s3_bucket.examplebucket_two, 0).id }"',
    ],
    ["parses splat expressions", '"${ toset(aws_s3_bucket.examplebucket.*) }"'],
    ["property access", `"\${ var.setting["bucket_name"] }"`],
    ["mixed property access", `"\${ var.setting["bucket_name"].arn }"`],
    ["variable property access", `"\${ var.settings[var.bucket_name_key] }"`],
    [
      "partial variable property access",
      `"\${ var.settings["\${var.prefix}-my-bucket"] }"`,
    ],
    ["list expressions", `"\${ [var.tags.app, var.tags.env] }"`],
  ])("%s", async (_, input) => {
    return expect(
      getExpressionAst("main.tf", input)
    ).resolves.toMatchSnapshot();
  });

  test("fails on malformed expressions", async () => {
    return expect(
      getExpressionAst("main.tf", '"${module.foo.output"')
    ).rejects.toMatchSnapshot();
  });
});
