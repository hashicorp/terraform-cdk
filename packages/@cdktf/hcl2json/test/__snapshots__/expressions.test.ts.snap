// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReferencesInExpression fails on malformed expressions 1`] = `[Error: :0,0-2: Unclosed template interpolation sequence; There is no closing brace for this interpolation sequence before the end of the file. This might be caused by incorrect nesting inside the given expression.]`;

exports[`getReferencesInExpression list expressions 1`] = `
Array [
  Object {
    "endPosition": 16,
    "startPosition": 4,
    "value": "var.tags.app",
  },
  Object {
    "endPosition": 30,
    "startPosition": 18,
    "value": "var.tags.env",
  },
]
`;

exports[`getReferencesInExpression mixed property access 1`] = `
Array [
  Object {
    "endPosition": 14,
    "startPosition": 3,
    "value": "var.setting",
  },
]
`;

exports[`getReferencesInExpression parses a reference string 1`] = `
Array [
  Object {
    "endPosition": 9,
    "startPosition": 2,
    "value": "var.foo",
  },
]
`;

exports[`getReferencesInExpression parses a simple string 1`] = `Array []`;

exports[`getReferencesInExpression parses a terraform function 1`] = `
Array [
  Object {
    "endPosition": 27,
    "startPosition": 10,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 41,
    "startPosition": 34,
    "value": "var.bar",
  },
]
`;

exports[`getReferencesInExpression parses embedded vars 1`] = `
Array [
  Object {
    "endPosition": 43,
    "startPosition": 26,
    "value": "module.foo.output",
  },
]
`;

exports[`getReferencesInExpression parses for in loops 1`] = `
Array [
  Object {
    "endPosition": 31,
    "startPosition": 22,
    "value": "var.users",
  },
]
`;

exports[`getReferencesInExpression parses four part references 1`] = `
Array [
  Object {
    "endPosition": 21,
    "startPosition": 2,
    "value": "data.type.name.attr",
  },
]
`;

exports[`getReferencesInExpression parses lists 1`] = `
Array [
  Object {
    "endPosition": 23,
    "startPosition": 6,
    "value": "module.foo.output",
  },
]
`;

exports[`getReferencesInExpression parses maps 1`] = `
Array [
  Object {
    "endPosition": 25,
    "startPosition": 8,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 80,
    "startPosition": 63,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 94,
    "startPosition": 87,
    "value": "var.bar",
  },
]
`;

exports[`getReferencesInExpression parses multi-line for loops 1`] = `
Array [
  Object {
    "endPosition": 43,
    "startPosition": 34,
    "value": "var.users",
  },
]
`;

exports[`getReferencesInExpression parses nested terraform functions 1`] = `
Array [
  Object {
    "endPosition": 21,
    "startPosition": 8,
    "value": "var.separator",
  },
  Object {
    "endPosition": 54,
    "startPosition": 37,
    "value": "module.foo.output",
  },
  Object {
    "endPosition": 68,
    "startPosition": 61,
    "value": "var.bar",
  },
]
`;

exports[`getReferencesInExpression parses splat expressions 1`] = `
Array [
  Object {
    "endPosition": 36,
    "startPosition": 9,
    "value": "aws_s3_bucket.examplebucket",
  },
]
`;

exports[`getReferencesInExpression parses ternary expressions 1`] = `
Array [
  Object {
    "endPosition": 48,
    "startPosition": 2,
    "value": "aws_kms_key.examplekms.deletion_window_in_days",
  },
  Object {
    "endPosition": 85,
    "startPosition": 55,
    "value": "aws_s3_bucket.examplebucket.id",
  },
]
`;

exports[`getReferencesInExpression parses terraform functions with file access 1`] = `
Array [
  Object {
    "endPosition": 42,
    "startPosition": 11,
    "value": "aws_s3_bucket.examplebucket_two",
  },
]
`;

exports[`getReferencesInExpression parses three part references 1`] = `
Array [
  Object {
    "endPosition": 19,
    "startPosition": 2,
    "value": "module.foo.output",
  },
]
`;

exports[`getReferencesInExpression parses variable arithmetics 1`] = `
Array [
  Object {
    "endPosition": 13,
    "startPosition": 2,
    "value": "var.members",
  },
  Object {
    "endPosition": 26,
    "startPosition": 16,
    "value": "var.admins",
  },
]
`;

exports[`getReferencesInExpression partial variable property access 1`] = `
Array [
  Object {
    "endPosition": 15,
    "startPosition": 3,
    "value": "var.settings",
  },
  Object {
    "endPosition": 29,
    "startPosition": 19,
    "value": "var.prefix",
  },
]
`;

exports[`getReferencesInExpression property access 1`] = `
Array [
  Object {
    "endPosition": 14,
    "startPosition": 3,
    "value": "var.setting",
  },
]
`;

exports[`getReferencesInExpression variable property access 1`] = `
Array [
  Object {
    "endPosition": 15,
    "startPosition": 3,
    "value": "var.settings",
  },
  Object {
    "endPosition": 35,
    "startPosition": 16,
    "value": "var.bucket_name_key",
  },
]
`;
